### PR TITLE
minio deploy  implement

### PR DESCRIPTION
**Pull request by soulseen**

### CODE REVIEW

File: api/v1/harborcluster_types.go [hunk 1]
```diff
@@ -164,7 +164,7 @@ type Oss struct {
 	Encrypt       string `json:"encrypt,omitempty"`
 	Secure        string `json:"secure,omitempty"`
 	Chunksize     string `json:"chunksize,omitempty"`
-	Rootdirectory string `json:"rootdirectory,omitempty"`
+	RootDirectory string `json:"rootdirectory,omitempty"`
```
Comment (wangcanfengxs) [inline api/v1/harborcluster_types.go:167]: Json struct tag needn't update?

File: controllers/storage/minio.go [hunk 2]
```diff
@@ -41,21 +51,30 @@ func (m *MinIOReconciler) Reconcile() (*lcm.CRStatus, error) {
 		return m.ProvisionExternalStorage()
 	}
 
-	err := m.KubeClient.Get(m.Ctx, m.getminIONamespacedName(), &minioCR)
+	err := m.KubeClient.Get(m.getminIONamespacedName(), &minioCR)
```
Comment (wangcanfengxs) [inline controllers/storage/minio.go:54]: method name `getminIONamespacedName` prefer to `getMinIONamespacedName`

File: controllers/storage/provision.go [hunk 3]
```diff
@@ -209,22 +249,30 @@ func (m *MinIOReconciler) generateSwiftSecret() *corev1.Secret {
...
-			"accesskeysecret": []byte(m.HarborCluster.Spec.Storage.Oss.Accesskeysecret),
-			"region":          []byte(m.HarborCluster.Spec.Storage.Oss.Region),
-			"bucket":          []byte(m.HarborCluster.Spec.Storage.Oss.Bucket),
-			"endpoint":        []byte(m.HarborCluster.Spec.Storage.Oss.Endpoint),
-			"internal":        []byte(m.HarborCluster.Spec.Storage.Oss.Internal),
-			"encrypt":         []byte(m.HarborCluster.Spec.Storage.Oss.Encrypt),
-			"secure":          []byte(m.HarborCluster.Spec.Storage.Oss.Secure),
-			"rootdirectory":   []byte(m.HarborCluster.Spec.Storage.Oss.Rootdirectory),
-			"chunksize":       []byte(m.HarborCluster.Spec.Storage.Oss.Chunksize)},
+			"authurl":             []byte(m.HarborCluster.Spec.Storage.Swift.Authurl),
```
Comment (wangcanfengxs) [inline controllers/storage/provision.go:252]: There are many fields didn't match `CamelCase`. Such as
`Authurl`, `Username`, `Tenantid`...

File: controllers/storage/minio.go [hunk 4]
```diff
@@ -41,21 +51,30 @@ func (m *MinIOReconciler) Reconcile() (*lcm.CRStatus, error) {
...
+	err = m.KubeClient.Get(m.getminIONamespacedName(), &minioStatefulSet)
+	if err != nil {
+		return minioNotReadyStatus(ErrorReason0, err.Error()), err
+	}
+
+	if minioStatefulSet.Status.ReadyReplicas == m.HarborCluster.Spec.Storage.InCluster.Spec.Replicas {
+		// TODO create default bucket
+		return m.ProvisionInClusterSecret(&minioCR)
+	}
 	return nil, nil
```
Comment (steven-zou) [inline controllers/storage/minio.go:71]: what does `nil, nil` mean here? Do nothing? Waiting for the next round reconcile process?

File: controllers/storage/provision.go [hunk 5]
```diff
@@ -1,36 +1,75 @@
...
+			Namespace:   m.HarborCluster.Namespace,
+			Labels:      m.getLabels(),
+			Annotations: m.generateAnnotations(),
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(minioInstamnce, HarborClusterMinIOGVK),
+			},
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			// TODO
```
Comment (steven-zou) [inline controllers/storage/provision.go:63]: What `TODO` needs to be done here?

File: controllers/storage/provision.go [hunk 6]
```diff
@@ -250,64 +298,76 @@ func (m *MinIOReconciler) generateOssSecret() *corev1.Secret {
...
 	//if err != nil {
 	//	return minioNotReadyStatus(ErrorReason2, err.Error()), err
 	//}
 	credsSecret := m.generateCredsSecret()
-	err := m.KubeClient.Create(m.Ctx, credsSecret)
+	err := m.KubeClient.Create(credsSecret)
 	if err != nil {
 		return minioNotReadyStatus(ErrorReason2, err.Error()), err
 	}
 	service := m.generateService()
```
Comment (steven-zou) [inline controllers/storage/provision.go:327]: For minIO, its CR is not a full accessible service? Why do you need to create service for minIO?

File: controllers/storage/provision.go [hunk 7]
```diff
@@ -139,16 +178,17 @@ func (m *MinIOReconciler) generateAzureSecret() *corev1.Secret {
 		},
 		Type: corev1.SecretTypeOpaque,
 		Data: map[string][]byte{
```
Comment (steven-zou) [inline controllers/storage/provision.go:180]: As secret is created, why do we still need to put the related info in the secret data field of the secret property?

File: controllers/storage/minio.go [hunk 8]
```diff
@@ -99,7 +118,12 @@ func minioReadyStatus(properties *lcm.Properties) *lcm.CRStatus {
 }
 
 func (m *MinIOReconciler) Delete() (*lcm.CRStatus, error) {
```
Comment (Ghostbaby) [inline controllers/storage/minio.go:120]: Maybe we don't need to delete the function? By setting the OwnerReference to let the controller-manager perform GC.

File: api/v1/harborcluster_types.go (hunk 1 above)
Comment (soulseen) [reply api/v1/harborcluster_types.go:167]: yes, all params are lowercase, ref [here](https://github.com/goharbor/harbor-operator/blob/c8e922bded7080598ee5d10861271b792a22086d/docs/installation/registry-storage-configuration.md#create-the-secret)

File: controllers/storage/provision.go (hunk 3 above)
Comment (soulseen) [reply controllers/storage/provision.go:252]: all params are lowercase in secret, ref [here](https://github.com/goharbor/harbor-operator/blob/c8e922bded7080598ee5d10861271b792a22086d/docs/installation/registry-storage-configuration.md#create-the-secret)

File: controllers/storage/minio.go (hunk 4 above)
Comment (soulseen) [reply controllers/storage/minio.go:71]: here mean minio is creating, continue Waiting for creating successful.

File: controllers/storage/provision.go (hunk 7 above)
Comment (soulseen) [reply controllers/storage/provision.go:180]: what do you meaning? 
the secret will be created after putting the related info in the secret, and harbor will get all property in the data field by secret name.

File: controllers/storage/provision.go (hunk 6 above)
Comment (soulseen) [reply controllers/storage/provision.go:327]: yes, we must add an external service to enable Service-based access to the MinIOInstance pods.

File: controllers/storage/minio.go (hunk 8 above)
Comment (soulseen) [reply controllers/storage/minio.go:120]: when the storage type switched, such as the type of inCluster change to external, we will use it to delete minio  CR.

File: api/v1/harborcluster_types.go (hunk 1 above)
Comment (wangcanfengxs) [reply api/v1/harborcluster_types.go:172]: In my opinion, This struct of HarborCluster is the API of ours, We should follow the go naming convention. And the secret is the resource that we need to create, but not related to the HarborCluster field naming.

File: controllers/storage/minio.go [hunk 9]
```diff
@@ -41,21 +51,30 @@ func (m *MinIOReconciler) Reconcile() (*lcm.CRStatus, error) {
...
-		// TODO need test
 		return m.Provision()
 	} else if err != nil {
 		return minioNotReadyStatus(ErrorReason0, err.Error()), err
 	}
 
+	var minioStatefulSet appsv1.StatefulSet
+	err = m.KubeClient.Get(m.getMinIONamespacedName(), &minioStatefulSet)
+	if err != nil {
+		return minioNotReadyStatus(ErrorReason0, err.Error()), err
```
Comment (wangcanfengxs) [inline controllers/storage/minio.go:64]: What does `ErrorReason0` means? The naming is too fuzzy.

File: controllers/storage/minio.go [hunk 10]
```diff
@@ -41,21 +51,30 @@ func (m *MinIOReconciler) Reconcile() (*lcm.CRStatus, error) {
...
-	err := m.KubeClient.Get(m.Ctx, m.getminIONamespacedName(), &minioCR)
+	err := m.KubeClient.Get(m.getMinIONamespacedName(), &minioCR)
 	if k8serror.IsNotFound(err) {
-		// TODO need test
 		return m.Provision()
 	} else if err != nil {
 		return minioNotReadyStatus(ErrorReason0, err.Error()), err
 	}
 
+	var minioStatefulSet appsv1.StatefulSet
```
Comment (wangcanfengxs) [inline controllers/storage/minio.go:61]: Why do you need to care about minioStatefulSet, but not minioCR?

File: controllers/storage/provision.go [hunk 11]
```diff
@@ -28,9 +25,50 @@ const (
...
+
+	p := &lcm.Property{
+		Name:  OssSecret,
+		Value: inClusterSecret.Name,
+	}
+	properties := &lcm.Properties{p}
+	return minioReadyStatus(properties), err
+}
+
+func (m *MinIOReconciler) generateInClusterSecret(minioInstamnce *minio.MinIOInstance) *corev1.Secret {
```
Comment (wangcanfengxs) [inline controllers/storage/provision.go:45]: This is a common method want to create secret？

Comment (soulseen) [reply controllers/storage/provision.go:45]: no, just for in-cluster secret

File: controllers/storage/minio.go (hunk 10 above)
Comment (soulseen) [reply controllers/storage/minio.go:61]: now the status in MinIO CR is not accurate, so we get status in minioStatefulSet to check the minIO is ready or not.

File: controllers/storage/minio.go (hunk 9 above)
Comment (soulseen) [reply controllers/storage/minio.go]: fixed

File: controllers/storage/minio.go (hunk 10 above)
Comment (wangcanfengxs) [reply controllers/storage/minio.go:61]: OK, I think you can use a method to check whether the minIO is ready or not. In the `Reconcile`, it does not need to care about the concrete workload(statefulset).

File: controllers/storage/minio.go [hunk 12]
```diff
@@ -41,21 +51,30 @@ func (m *MinIOReconciler) Reconcile() (*lcm.CRStatus, error) {
...
+
+	var minioStatefulSet appsv1.StatefulSet
+	err = m.KubeClient.Get(m.getMinIONamespacedName(), &minioStatefulSet)
+	if err != nil {
+		return minioNotReadyStatus(GetMinIOError, err.Error()), err
 	}
 
+	if minioStatefulSet.Status.ReadyReplicas == m.HarborCluster.Spec.Storage.InCluster.Spec.Replicas {
+		// TODO create default bucket
+		return m.ProvisionInClusterSecretAsOss(&minioCR)
```
Comment (wangcanfengxs) [inline controllers/storage/minio.go:69]: What does `ProvisionInClusterSecretAsOss` means? Why `AsOss`?

File: api/v1/harborcluster_types.go (hunk 1 above)
Comment (soulseen) [reply api/v1/harborcluster_types.go:172]: https://docs.docker.com/registry/configuration/#storage all storage field are lowercase, I think we should follow this standard.

File: controllers/storage/minio.go (hunk 12 above)
Comment (soulseen) [reply controllers/storage/minio.go:69]: minIO implements both interfaces of Oss and S3, so we can create in-cluster secret as Oss or s3.

**Pull request merged**